test/quantization/fx/test_quantize_fx.py

```diff
@@ -5917,7 +5917,8 @@ def forward(self, x):
             ns.call_function(torch.quantize_per_tensor): 1,
             ns.call_function(torch.ops.quantized.linear): 1,
             ns.call_function(torch.ops.quantized.layer_norm): 1,
-            ns.call_method("dequantize"): 1,
+            ns.call_function(torch.quantize_per_channel): 1,
```

Comment:
can you paste the full model before and after the change? would like to see if the final model makes sense or not

Reply:
Before the change:
```python
quant type:                                                                                                                                                                                               [3/1820]
 QuantType.STATIC                                                                                                                                                                                                 
original model:                                                                                                                                                                                                   
 M(                                                                                                                                                                                                               
  (mods1): Linear()                                                                                                                                                                                               
)                                                                                                                                                                                                                 
                                                                                                                                                                                                                  
prepared model:                                                                                                                                                                                                   
 GraphModule(                                                                                                                                                                                                     
  (activation_post_process_0): HistogramObserver()                                                                                                                                                                
  (mods1): Module()                                                                                                                                                                                               
  (activation_post_process_1): PerChannelMinMaxObserver(min_val=tensor([1., 1., 1., 1., 1.]), max_val=tensor([1., 1., 1., 1., 1.]))                                                                               
  (activation_post_process_2): HistogramObserver()                                                                                                                                                                
  (activation_post_process_3): PerChannelMinMaxObserver(min_val=tensor([-0.9331, -1.3769, -1.0491, -0.2322, -1.0875]), max_val=tensor([0.3867, 1.6512, 2.4418, 0.7138, 1.4836]))                                  
  (activation_post_process_4): HistogramObserver()                                                                                                                                                                
)                                                                                                                                                                                                                 
                                                                                                                                                                                                                  
                                                                                                                                                                                                                  
                                                                                                                                                                                                                  
def forward(self, x):                                                                                                                                                                                             
    activation_post_process_0 = self.activation_post_process_0(x);  x = None                                                                                                                                      
    mods1_w = self.mods1.w                                                                                                                                                                                        
    activation_post_process_1 = self.activation_post_process_1(mods1_w);  mods1_w = None                                                                                                                          
    mods1_b = self.mods1.b                                                                                                                                                                                        
    linear = torch._C._nn.linear(activation_post_process_0, activation_post_process_1, mods1_b);  activation_post_process_0 = activation_post_process_1 = mods1_b = None                                          
    activation_post_process_2 = self.activation_post_process_2(linear);  linear = None                                                                                                                            
    scale = self.scale                                                                                                                                                                                            
    activation_post_process_3 = self.activation_post_process_3(scale);  scale = None                                                                                                                              
    bias = self.bias                                                                                                                                                                                              
    layer_norm = torch.nn.functional.layer_norm(activation_post_process_2, [5, 5], weight = activation_post_process_3, bias = bias, eps = 1e-05);  activation_post_process_2 = activation_post_process_3 = bias = 
None                                                                                                                                                                                                              
    activation_post_process_4 = self.activation_post_process_4(layer_norm);  layer_norm = None                                                                                                                           "devvm5875.lla0.facebo" 08:43 30-Aug-22
    return activation_post_process_4
     
quantized model:
 GraphModule()



def forward(self, x):
    mods1_input_scale_0 = self.mods1_input_scale_0
    mods1_input_zero_point_0 = self.mods1_input_zero_point_0
    quantize_per_tensor = torch.quantize_per_tensor(x, mods1_input_scale_0, mods1_input_zero_point_0, torch.quint8);  x = mods1_input_scale_0 = mods1_input_zero_point_0 = None
    mods1_packed_weight_0 = self.mods1_packed_weight_0
    mods1_scale_1 = self.mods1_scale_1
    mods1_zero_point_1 = self.mods1_zero_point_1
    linear = torch.ops.quantized.linear(quantize_per_tensor, mods1_packed_weight_0, mods1_scale_1, mods1_zero_point_1);  quantize_per_tensor = mods1_packed_weight_0 = mods1_scale_1 = mods1_zero_point_1 = None
    scale = self.scale
    bias = self.bias
    _scale_1 = self._scale_1
    _zero_point_1 = self._zero_point_1
    layer_norm_1 = torch.ops.quantized.layer_norm(linear, [5, 5], weight = scale, bias = bias, eps = 1e-05, output_scale = _scale_1, output_zero_point = _zero_point_1);  linear = scale = bias = _scale_1 = _zero
_point_1 = None
    dequantize_4 = layer_norm_1.dequantize();  layer_norm_1 = None
    return dequantize_4
 ```

Reply:
After the change:
```python
quant type:                                                                                                                                                                                                                                                          
 QuantType.STATIC                                                                                                                                                                                                                                                    
original model:                                                                                                                                                                                                                                                      
 M(                                                                                                                                                                                                                                                                  
  (mods1): Linear()                                                                                                                                                                                                                                                  
)                                                                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                                     
prepared model:                                                                                                                                                                                                                                                      
 GraphModule(                                                                                                                                                                                                                                                        
  (activation_post_process_0): HistogramObserver()                                                                                                                                                                                                                   
  (mods1): Module()                                                                                                                                                                                                                                                  
  (activation_post_process_1): PerChannelMinMaxObserver(min_val=tensor([1., 1., 1., 1., 1.]), max_val=tensor([1., 1., 1., 1., 1.]))                                                                                                                                  
  (activation_post_process_2): HistogramObserver()                                                                                                                                                                                                                   
  (activation_post_process_3): PerChannelMinMaxObserver(min_val=tensor([-0.9331, -1.3769, -1.0491, -0.2322, -1.0875]), max_val=tensor([0.3867, 1.6512, 2.4418, 0.7138, 1.4836]))                                                                                     
  (activation_post_process_4): HistogramObserver()                                                                                                                                                                                                                   
)                                                                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                     
def forward(self, x):                                                                                                                                                                                                                                                
    activation_post_process_0 = self.activation_post_process_0(x);  x = None                                                                                                                                                                                         
    mods1_w = self.mods1.w                                                                                                                                                                                                                                           
    activation_post_process_1 = self.activation_post_process_1(mods1_w);  mods1_w = None                                                                                                                                                                             
    mods1_b = self.mods1.b                                                                                                                                                                                                                                           
    linear = torch._C._nn.linear(activation_post_process_0, activation_post_process_1, mods1_b);  activation_post_process_0 = activation_post_process_1 = mods1_b = None                                                                                             
    activation_post_process_2 = self.activation_post_process_2(linear);  linear = None                                                                                                                                                                               
    scale = self.scale                                                                                                                                                                                                                                               
    activation_post_process_3 = self.activation_post_process_3(scale);  scale = None                                                                                                                                                                                 
    bias = self.bias                                                                                                                                                                                                                                                 
    layer_norm = torch.nn.functional.layer_norm(activation_post_process_2, [5, 5], weight = activation_post_process_3, bias = bias, eps = 1e-05);  activation_post_process_2 = activation_post_process_3 = bias = None                                               
    activation_post_process_4 = self.activation_post_process_4(layer_norm);  layer_norm = None                                                                                                                                                                       
    return activation_post_process_4                                                                                                                                                                                                                                 
                                                                                                                                                                                                                                                                                                                                                                                                                                 
                                                                                                                                                                                                                                                                     
quantized model:                                                                                                                                                                                                                                                     
 GraphModule()                                                                                                                                                                                                                                                       
                                                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                     
def forward(self, x):                                                                                                                                                                                                                                                
    mods1_input_scale_0 = self.mods1_input_scale_0                                                                                                                                                                                                                   
    mods1_input_zero_point_0 = self.mods1_input_zero_point_0                                                                                                                                                                                                         
    quantize_per_tensor = torch.quantize_per_tensor(x, mods1_input_scale_0, mods1_input_zero_point_0, torch.quint8);  x = mods1_input_scale_0 = mods1_input_zero_point_0 = None                                                                                      
    mods1_packed_weight_0 = self.mods1_packed_weight_0                                                                                                                                                                                                               
    mods1_scale_1 = self.mods1_scale_1                                                                                                                                                                                                                               
    mods1_zero_point_1 = self.mods1_zero_point_1                                                                                                                                                                                                                     
    linear = torch.ops.quantized.linear(quantize_per_tensor, mods1_packed_weight_0, mods1_scale_1, mods1_zero_point_1);  quantize_per_tensor = mods1_packed_weight_0 = mods1_scale_1 = mods1_zero_point_1 = None                                                     
    scale = self.scale                                                                                                                                                                                                                                               
    _scale_0 = self._scale_0                                                                                                                                                                                                                                         
    _zero_point_0 = self._zero_point_0                                                                                                                                                                                                                               
    quantize_per_channel_1 = torch.quantize_per_channel(scale, _scale_0, _zero_point_0, 0, torch.qint8);  scale = _scale_0 = _zero_point_0 = None                                                                                                                    
    dequantize_3 = quantize_per_channel_1.dequantize();  quantize_per_channel_1 = None                                                                                                                                                                               
    bias = self.bias                                                                                                                                                                                                                                                 
    _scale_1 = self._scale_1                                                                                                                                                                                                                                         
    _zero_point_1 = self._zero_point_1                                                                                                                                                                                                                               
    layer_norm_1 = torch.ops.quantized.layer_norm(linear, [5, 5], weight = dequantize_3, bias = bias, eps = 1e-05, output_scale = _scale_1, output_zero_point = _zero_point_1);  linear = dequantize_3 = bias = _scale_1 = _zero_point_1 = None                      
    dequantize_4 = layer_norm_1.dequantize();  layer_norm_1 = None                                                                                                                                                                                                   
    return dequantize_4   
```

Reply:
I think there are some problems with the lowering code for quantized layer_norm, we need to fix that, basically the lowering does did not fuse the dequantize into the quantized layer_norm op (only activation was fused), the fusion pattern should be:
```
weight - quant - dequant \
input - quant - dequant - layer_norm - quant - dequant
```

Reply:
> Gotcha, are you talking about dequantize_3?

Yes.

> Is the expected behavior to be able to pass in the quantized scale directly to quantized.layer_norm?

I guess the weight of layernorm is not quantized: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp#L2672

that means we need to skip adding q/dq for weight, this requires changes to both the qconfig for layer_norm and the BackendConfig entry for layernorm

Reply:
@jerryzh168 Does this look right?

```python

quant type:
 QuantType.STATIC
original model:
 M(
  (mods1): Linear()
)

prepared model:
 GraphModule(
  (activation_post_process_0): HistogramObserver()
  (mods1): Module()
  (activation_post_process_1): PerChannelMinMaxObserver(min_val=tensor([1., 1., 1., 1., 1.]), max_val=tensor([1., 1., 1., 1., 1.]))
  (activation_post_process_2): HistogramObserver()
  (activation_post_process_3): PlaceholderObserver()
  (activation_post_process_4): HistogramObserver()
)



def forward(self, x):
    activation_post_process_0 = self.activation_post_process_0(x);  x = None
    mods1_w = self.mods1.w
    activation_post_process_1 = self.activation_post_process_1(mods1_w);  mods1_w = None
    mods1_b = self.mods1.b
    linear = torch._C._nn.linear(activation_post_process_0, activation_post_process_1, mods1_b);  activation_post_process_0 = activation_post_process_1 = mods1_b = None
    activation_post_process_2 = self.activation_post_process_2(linear);  linear = None
    scale = self.scale
    activation_post_process_3 = self.activation_post_process_3(scale);  scale = None
    bias = self.bias
    layer_norm = torch.nn.functional.layer_norm(activation_post_process_2, [5, 5], weight = activation_post_process_3, bias = bias, eps = 1e-05);  activation_post_process_2 = activation_post_process_3 = bias = None
    activation_post_process_4 = self.activation_post_process_4(layer_norm);  layer_norm = None
    return activation_post_process_4
    
# To see more debug info, please use `graph_module.print_readable()`

quantized model:
 GraphModule()



def forward(self, x):
    mods1_input_scale_0 = self.mods1_input_scale_0
    mods1_input_zero_point_0 = self.mods1_input_zero_point_0
    quantize_per_tensor = torch.quantize_per_tensor(x, mods1_input_scale_0, mods1_input_zero_point_0, torch.quint8);  x = mods1_input_scale_0 = mods1_input_zero_point_0 = None
    mods1_packed_weight_0 = self.mods1_packed_weight_0
    mods1_scale_1 = self.mods1_scale_1
    mods1_zero_point_1 = self.mods1_zero_point_1
    linear = torch.ops.quantized.linear(quantize_per_tensor, mods1_packed_weight_0, mods1_scale_1, mods1_zero_point_1);  quantize_per_tensor = mods1_packed_weight_0 = mods1_scale_1 = mods1_zero_point_1 = None
    scale = self.scale
    bias = self.bias
    _scale_0 = self._scale_0
    _zero_point_0 = self._zero_point_0
    layer_norm_1 = torch.ops.quantized.layer_norm(linear, [5, 5], weight = scale, bias = bias, eps = 1e-05, output_scale = _scale_0, output_zero_point = _zero_point_0);  linear = scale = bias = _scale_0 = _zero_point_0 = None
    dequantize_3 = layer_norm_1.dequantize();  layer_norm_1 = None
    return dequantize_3
    
# To see more debug info, please use `graph_module.print_readable()`
'placeholder' 'x' 'x' () {}
'get_attr' 'mods1_input_scale_0' 'mods1_input_scale_0' () {}
'get_attr' 'mods1_input_zero_point_0' 'mods1_input_zero_point_0' () {}
'call_function' 'quantize_per_tensor' <built-in method quantize_per_tensor of type object at 0x7f9ea0a13180> (x, mods1_input_scale_0, mods1_input_zero_point_0, torch.quint8) {}
'get_attr' 'mods1_packed_weight_0' 'mods1_packed_weight_0' () {}
'get_attr' 'mods1_scale_1' 'mods1_scale_1' () {}
'get_attr' 'mods1_zero_point_1' 'mods1_zero_point_1' () {}
'call_function' 'linear' <OpOverloadPacket(op='quantized.linear')> (quantize_per_tensor, mods1_packed_weight_0, mods1_scale_1, mods1_zero_point_1) {}
'get_attr' 'scale' 'scale' () {}
'get_attr' 'bias' 'bias' () {}
'get_attr' '_scale_0' '_scale_0' () {}
'get_attr' '_zero_point_0' '_zero_point_0' () {}
'call_function' 'layer_norm_1' <OpOverloadPacket(op='quantized.layer_norm')> (linear, [5, 5]) {'weight': scale, 'bias': bias, 'eps': 1e-05, 'output_scale': _scale_0, 'output_zero_point': _zero_point_0}
'call_method' 'dequantize_3' 'dequantize' (layer_norm_1,) {}
'output' 'output' 'output' (dequantize_3,) {}
```

Reply:
yeah this looks good